File: medium/1004. Max Consecutive Ones III/1004. Max Consecutive Ones III.py
from typing import List


class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        remind = k
        count = 0

        n = len(nums)
        res = count
        left = 0

        for right in range(n):
            if nums[right] == 1:
                count += 1
            else:
                if remind:
                    remind -= 1
                    count += 1
                else:
                    while not remind:
                        if nums[left] == 0:
                            remind += 1
                        left += 1
                        count -= 1

                    remind -= 1
                    count += 1

            res = max(res, count)

        return res

File: medium/1004. Max Consecutive Ones III/1004. Max Consecutive Ones III.go
package main

func longestOnes(nums []int, k int) int {
	count := 0
	res := 0
	left := 0
	remind := k

	for right := 0; right < len(nums); right++ {
		if nums[right] == 0 {
			count += 1
		} else {
			if remind != 0 {
				remind -= 1
				count += 1
			} else {
				for remind == 0 {
					if nums[left] == 0 {
						remind = 1
					}
					left += 1
					count -= 1
				}

				remind -= 1
				count += 1
			}
		}

		res = max(res, count)
	}

	return res
}
